[Multilane] Provides lane-to-lane methods to create Connections from the Builder

agalbachicar opened the pull request:
As part of issue #7194, and parent issue #4934, this PR includes:

- New constructors to `Connection` class so as to derive reference curve from different lane information (`Endpoint` and `EndpointZ`).
- `Builder` provides new methods to construct Arc and Line `Connections` from other connections.
- Modifies `Builder` tests to create the same `RoadGeometry` with both reference curve and lane-to-lane API.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7471)
<!-- Reviewable:end -->

agalbachicar commented:
+@maddog-tri for feature review please.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-:-KyvlIiDI4spU8wemImO:b-5klkw7)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
First pass complete, I think --- some interface changes are needed, so I haven't gone deep into the test code yet. PTAL.

---

Reviewed 3 of 6 files at r1.
Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9lQY-PNRwFo6j5qG2:-Kz9lQY-PNRwFo6j5qG3:b20mtuy) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/builder.cc#L59)):*
> ```cpp
>                                    int end_lane_index, const EndpointZ& z_end) {
>   DRAKE_DEMAND(start_lane_index < num_lanes && start_lane_index >= 0);
>   DRAKE_DEMAND(end_lane_index < num_lanes && start_lane_index >= 0);
> ```

`start_lane_index` should be `end_lane_index`.

---

*[drake/automotive/maliput/multilane/builder.cc, line 81 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9m-RkX4pFYSLWulnl:-Kz9m-RkX4pFYSLWulnm:b6v4dk) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/builder.cc#L81)):*
> ```cpp
>                                    int end_lane_index, const EndpointZ& z_end) {
>   DRAKE_DEMAND(start_lane_index < num_lanes && start_lane_index >= 0);
>   DRAKE_DEMAND(end_lane_index < num_lanes);
> ```

Doesn't assert `>= 0` on `end_lane_index`.

---

*[drake/automotive/maliput/multilane/builder.h, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9iOZSuEEhN7V9dbUZ:-Kz9iOZTi4GuX0NZYuzU:b-7m7he5) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/builder.h#L124)):*
> ```objc
>                             int conn_start_lane_index, bool is_start_endpoint,
>                             double length, int end_lane_index,
>                             const EndpointZ& z_end);
> ```

For these two new `Connect()` methods, it would be more versatile to match the corresponding `Connection()` constructors:  replace `start_connection`, `conn_start_lane_index`,  and `is_start_endpoint` with a single `const Endpoint& start` parameter.

The basic lane-to-lane usage would look like:
```
  auto c2 = b.Connect("c2", 5, kLeftShoulder, kRightShoulder, c1->LaneEnd(3), kLength, 3, kEndZ);
```
but it would also allow initiating a connection from a bare Endpoint:
```
  const Endpoint start{{...}, {...}};
  auto c1 = b.Connect("c1", 5, kLeftShoulder, kRightShoulder, start, kLength, 3, kEndZ);
```
And, it improves the structure of the arguments at the call-sites, too.  :)

---

*[drake/automotive/maliput/multilane/connection.h, line 246 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9oPqK6pNHBAH0yCKS:-Kz9oPqLL5J6qXhFVPqD:b-uipix3) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/connection.h#L246)):*
This contract (and the next method's) is not quite right, and I misspoke (miswrote?) about it in our last slack-conversation when I wrote "yes, [...] line/arc arguments are relative to the start Endpoint, and control the lane associated with the start Endpoint."

Going back to #7194:
> * given reference-line at offset `r_ref` relative to lane `n_ref`
> * given a reference-line shape specified by a displacement which is either:
>   * linear, specified by a length L, or
>   * arc, specified (via an ArcOffset) by an arc-length phi and a radius R of the center from the reference-line

The line/arc arguments do control the geometry of the reference-curve, but the reference curve is not necessarily coincident with the `start_lane_index` lane.  Instead, the origin of the reference is offset by `r_ref` (along the the superelevated road surface) from the `n_ref` lane (which is not necessarily `start_lane_index`!).

This extra degree of freedom corresponds to `r0` in the original two `Connection()` constructors, and it is necessary for the same reason here:  the reference line defines the origin of the superelevation rotation, which might (often) need to be somewhere other than the centerline of a lane.

---

*[drake/automotive/maliput/multilane/connection.h, line 253 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9yLpaKKu8O-cSPZeg:-Kz9yLpaKKu8O-cSPZeh:byl85rb) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/connection.h#L253)):*
> ```objc
>   /// First Lane centerline will be placed at `start_lane_index` times
>   /// `lane_width` distance from the reference curve. Each lane's width will be
>   /// `lane_width`, which should be greater or equal to zero.
> ```

(In lieu of the above contract change, this paragraph will need tweaking, too.)

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 507 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-KzA-PWdVwEwfCVSmbVJ:-KzA-PWdVwEwfCVSmbVK:b4rt2f8) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L507)):*
> ```cpp
>   EXPECT_EQ(rg->id(), api::RoadGeometryId("multilane-cross"));
> 
>   EXPECT_EQ(rg->num_junctions(), junction_truth_map.size());
> ```

This check is missing in the `LaneToLaneMultilaneCross` version below.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 508 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-KzA-_bly3QeTDXSVLRn:-KzA-_bly3QeTDXSVLRo:b-yeiz8s) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L508)):*
> ```cpp
> 
>   EXPECT_EQ(rg->num_junctions(), junction_truth_map.size());
>   for (int i = 0; i < rg->num_junctions(); i++) {
> ```

There _must_ be a nice way to not duplicate all this code below, maybe something with parameterized tests, or at the very least stick it in a method in the `MultilaneBuilderCrossTest` class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-:-KzAAhBOBYPWpBRno8zN:bum47es)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Thanks for the review and sorry for the delay. I've addressed all the comments. PTAL. 

---

Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9lQY-PNRwFo6j5qG2:-KzAgEItJzZQHUW6am8P:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/builder.cc#L59)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

`start_lane_index` should be `end_lane_index`.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.cc, line 81 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9m-RkX4pFYSLWulnl:-KzAgDoCaP6X3cl8s-jz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/builder.cc#L81)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Doesn't assert `>= 0` on `end_lane_index`.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9iOZSuEEhN7V9dbUZ:-KzAazmGCk3X9L6ZvWjr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/builder.h#L124)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

For these two new `Connect()` methods, it would be more versatile to match the corresponding `Connection()` constructors:  replace `start_connection`, `conn_start_lane_index`,  and `is_start_endpoint` with a single `const Endpoint& start` parameter.

The basic lane-to-lane usage would look like:
```
  auto c2 = b.Connect("c2", 5, kLeftShoulder, kRightShoulder, c1->LaneEnd(3), kLength, 3, kEndZ);
```
but it would also allow initiating a connection from a bare Endpoint:
```
  const Endpoint start{{...}, {...}};
  auto c1 = b.Connect("c1", 5, kLeftShoulder, kRightShoulder, start, kLength, 3, kEndZ);
```
And, it improves the structure of the arguments at the call-sites, too.  :)

</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/connection.h, line 246 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9oPqK6pNHBAH0yCKS:-KziHkVzaI4jyji7llEu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/connection.h#L246)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

This contract (and the next method's) is not quite right, and I misspoke (miswrote?) about it in our last slack-conversation when I wrote "yes, [...] line/arc arguments are relative to the start Endpoint, and control the lane associated with the start Endpoint."

Going back to #7194:
> * given reference-line at offset `r_ref` relative to lane `n_ref`
> * given a reference-line shape specified by a displacement which is either:
>   * linear, specified by a length L, or
>   * arc, specified (via an ArcOffset) by an arc-length phi and a radius R of the center from the reference-line

The line/arc arguments do control the geometry of the reference-curve, but the reference curve is not necessarily coincident with the `start_lane_index` lane.  Instead, the origin of the reference is offset by `r_ref` (along the the superelevated road surface) from the `n_ref` lane (which is not necessarily `start_lane_index`!).

This extra degree of freedom corresponds to `r0` in the original two `Connection()` constructors, and it is necessary for the same reason here:  the reference line defines the origin of the superelevation rotation, which might (often) need to be somewhere other than the centerline of a lane.


</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/connection.h, line 253 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-Kz9yLpaKKu8O-cSPZeg:-KziHl3nOw0Jhr_IRQKt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/connection.h#L253)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

(In lieu of the above contract change, this paragraph will need tweaking, too.)
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 507 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-KzA-PWdVwEwfCVSmbVJ:-KzTc7mgoWk2JeS60NsP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L507)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

This check is missing in the `LaneToLaneMultilaneCross` version below.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 508 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-KzA-_bly3QeTDXSVLRn:-KzTc851lgDvke6am69B:bafxrl9) ([raw file](https://github.com/robotlocomotion/drake/blob/2c60a764af86d9e8b05cc9c420986741524bd582/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L508)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

There _must_ be a nice way to not duplicate all this code below, maybe something with parameterized tests, or at the very least stick it in a method in the `MultilaneBuilderCrossTest` class.
</blockquote></details>

Done. I checked for a GTest idiomatic expression to represent this but I couldn't find one. I chose to move it to a function in the test class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-:-KzioiZuCpI2-xt7Q4mU:b-tsdhvi)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Likewise.  The "endpoint translation" math is actually quite tricky; see comments below.

---

Reviewed 3 of 6 files at r2.
Review status: 3 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-E3YDM-CHupNKgu9jD:-L-E3YDM-CHupNKgu9jE:bre2bal) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/builder.cc#L61)):*
> ```cpp
>   connections_.push_back(std::make_unique<Connection>(
>       id, start, start_lane_index, z_end, end_lane_index, num_lanes, 0.,
>       start_lane_index, lane_width_, left_shoulder, right_shoulder, length));
> ```

`r_ref` and `lane_ref_index` need to be exposed as parameters to `Connect(...)` (here and the next method below).

It they are hard-coded constants in the only place that the `Connection()` constructor is called, there is not much point in having them as free parameters in `Connection()`.  ;^)

---

*[drake/automotive/maliput/multilane/builder.h, line 103 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-GBxg2xfY0AX31AQ90:-L-GBxg2xfY0AX31AQ91:bscav6s) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/builder.h#L103)):*
> ```objc
>   /// Creates a line Connection whose ID is `id` and has `num_lanes` lanes.
>   ///
>   /// Line length is `length`.
> ```

More specifically, `length` specifies the length of the reference curve.

---

*[drake/automotive/maliput/multilane/builder.h, line 122 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-GB9B9SMKwc9aYcG_8:-L-GB9B9SMKwc9aYcG_9:bm4j9nk) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/builder.h#L122)):*
> ```objc
>   /// Creates an arc Connection whose ID is `id` and has `num_lanes` lanes.
>   ///
>   /// `arc` specifies the shape of connection's `start_lane_index` lane.
> ```

Actually, the shape of the reference curve (which will cease being the same as the `start_lane_index` lane when the lane_ref_index and r_ref parameters are available).

---

*[drake/automotive/maliput/multilane/connection.cc, line 136 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-G5FxctxRnfSTZFX3Y:-L-G5FxctxRnfSTZFX3Z:b394oc8) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/connection.cc#L136)):*
> ```cpp
>   // Start and end EndpointXys should be translated to the reference curve
>   // position. In order to compute new xy positions, superelevation at the
>   // extents must be taken into account.
> ```

The code below doesn't actually take superelevation into account for computing xy positions.

Also... it's even trickier than that, because non-zero superelevation also causes `z'` to couple into xy and z positions, and `theta_dot` couples into `heading`, and....

What really needs to happen is to take the equations from `RoadCurve::W_of_prh()` and `RoadCurve::Orientation()` and "solve" them for the endpoint parameters.  In particular:
  * `start` is at `(0, r_start_lane, 0)` in segment-frame.... so the way to calculate `start_` below is to figure out what it needs to be such that `W_of_prh(0, r_start_lane, 0)` and `Orientation(0, r_start_lane, 0)` yield answers consistent with `start`.
  * Given `start_`, the XY component of `end_` is derived by extending `start_` by `line_length` (or `arc_offset` for the arc version).
  * `end_z` occurs at `(p,r,h) = (1, r_end_lane, 0)`.  Given that and `end_.xy()`, `end_.z()` needs to be derived by solving `W_of_prh()` and `Orientation()` again.

I have not cranked through the equations for all this yet, though at some point I did convince myself that finding the solutions is feasible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-:-L-GH7wDYKtTU29hfUKC:bizfrl5)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
*[drake/automotive/maliput/multilane/connection.cc, line 136 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-G5FxctxRnfSTZFX3Y:-L-Xt0qMcPdpB0sGAwFX:bemd9b) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/connection.cc#L136)):*
> e not cranked through the equations for all this yet, though at some point I did convince myself that finding the solutions i

I knew that this code did not completely support superelevation and I want to apologize for not providing an explicit comment about it.

I guess that the proposed solution and its test suite would involve such an effort that may be out of the scope of this PR, or even exceed the LOC constraint... I wonder if it'd be better to defer the complete solution to another PR with a support issue where we can discuss all the math details and then a PR that addresses them. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
*[drake/automotive/maliput/multilane/connection.cc, line 136 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-G5FxctxRnfSTZFX3Y:-L-bjUQAUatVDpdtDvuN:bghyt3t) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/connection.cc#L136)):*
<details><summary><i>Previously, agalbachicar (Agustin Alba Chicar) wrote…</i></summary><blockquote>

> e not cranked through the equations for all this yet, though at some point I did convince myself that finding the solutions i

I knew that this code did not completely support superelevation and I want to apologize for not providing an explicit comment about it.

I guess that the proposed solution and its test suite would involve such an effort that may be out of the scope of this PR, or even exceed the LOC constraint... I wonder if it'd be better to defer the complete solution to another PR with a support issue where we can discuss all the math details and then a PR that addresses them. What do you think?
</blockquote></details>

Ah-ha --- in that case, we need a `TODO` that explains what works/doesn't-work in this version of the code. I'm ok pushing the fix into a subsequent PR, but this needs to get sorted out before any work on the YAML interface, because it is possible that something about this Connection interface has to change in order to make this correct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
After the required change on the Builder API, I found an issue so it took me some time to fix it and provide the appropriate test coverage. I think that all the supported geometries are tested now and the API is exercised in different ways. However, I am not totally satisfied with the result. You will find the same pattern in the complex multi-lane intersection test  with the single segment geometries, entity creation differs but code that test is the same. Any suggestion is welcome.

Soon I'll open the issue to discuss how to include superelevation to the equations with a draft, questions, etc.

PTAL.

---

Review status: 1 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-E3YDM-CHupNKgu9jD:-L-cH0X658MXa6ZNtpRg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/builder.cc#L61)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

`r_ref` and `lane_ref_index` need to be exposed as parameters to `Connect(...)` (here and the next method below).

It they are hard-coded constants in the only place that the `Connection()` constructor is called, there is not much point in having them as free parameters in `Connection()`.  ;^)
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 103 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-GBxg2xfY0AX31AQ90:-L-X5E1cOvX_Hxi26rHp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/builder.h#L103)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

More specifically, `length` specifies the length of the reference curve.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 122 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-GB9B9SMKwc9aYcG_8:-L-X5KySMVnp2H5gqQD8:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/builder.h#L122)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Actually, the shape of the reference curve (which will cease being the same as the `start_lane_index` lane when the lane_ref_index and r_ref parameters are available).
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/connection.cc, line 136 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-L-G5FxctxRnfSTZFX3Y:-L-hc2FX0Z89lUl9lpL1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a7b97c365d9bb153dab108cbc2f4d77ca1f6520d/drake/automotive/maliput/multilane/connection.cc#L136)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Ah-ha --- in that case, we need a `TODO` that explains what works/doesn't-work in this version of the code. I'm ok pushing the fix into a subsequent PR, but this needs to get sorted out before any work on the YAML interface, because it is possible that something about this Connection interface has to change in order to make this correct.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-:-L-hc4KzFKM55cjFHBV1:bmqwrz8)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
I've made an update of this PR with the folder changes in introduced in #7575

---

Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-:-L0QnJ_aIlnZ2Hve9lZQ:b-q81eia)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Could you close this PR and re-open (with the current state of the code) as a fresh PR, with a reference in the commit msg to this original PR.

I had horrible problems with Reviewable getting very confused by the #7575 changes in another PR; I don't want to risk going through that again.  I will go and check the new PR tree against the comment threads in this PR after you do that.

---

Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-:-L5GP8VMf4oG9ENEGAgT:b-lh2ih1)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
I'm closing this PR as @maddog-tri requested in favor of a fresh one with recent master changes.

agalbachicar commented:
Follow up PR: #8077 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7471#-:-L5OiGjAZIStcldYAN_V:bk0z2rs)*
<!-- Sent from Reviewable.io -->
